source/CVODEIntegrator.cpp

```diff
@@ -281,10 +324,188 @@ namespace rr
 		return Integrator::Deterministic;
 	}
 
+	void CVODEIntegrator::setIndividualTolerance(int index, double value) {
+		// check if index is valid
+		checkIndex(index, mModel->getNumFloatingSpecies());
+
+		// the tolerance vector that will be stored
+		vector<double> v;
+		
+		switch (getType("absolute_tolerance")) {
+		
+			// all cases below could be convert to a double type
+			case Variant::TypeId::INT32:
+			case Variant::TypeId::INT64:
+			case Variant::TypeId::UINT32:
+			case Variant::TypeId::UINT64:
+			case Variant::TypeId::FLOAT:
+			case Variant::TypeId::DOUBLE:
+			{
+				// scalar tolerance
+				// need to be converted to vector tolerance since tolerance of individual species is set
+				double abstol = CVODEIntegrator::getValueAsDouble("absolute_tolerance");
+				for (int i = 0; i < mModel->getNumFloatingSpecies(); i++)
+					v.push_back(i == index ? value : abstol);
+				break;
+			}
+
+			case Variant::TypeId::DOUBLEVECTOR:
+			{
+				// vector tolerance
+				v = CVODEIntegrator::getValueAsDoubleVector("absolute_tolerance");
+				// only need to update the corresponding species
+				v[index] = value;
+				break;
+			}
+
+			default:
+				throw std::runtime_error("CVODEIntegrator::setIndividualTolerance failed, double or double vector expected");
+				break;
+		}
+
+		// update the tolerance
+		CVODEIntegrator::setValue("absolute_tolerance", v);
+	}
+
+
+	void CVODEIntegrator::setConcentrationTolerance(const Variant& value) {
+
+		uint ncomp = mModel->getNumCompartments();
+
+		double* volumes = (double*)calloc(ncomp, sizeof(double));
+		mModel->getCompartmentVolumes(ncomp, 0, volumes);
+
+		// the tolerance vector that will be stored
+		vector<double> v;
+
+		switch (value.type()) {
+
+			// all cases below could be convert to a double type
+			case Variant::TypeId::INT32:
+			case Variant::TypeId::INT64:
+			case Variant::TypeId::UINT32:
+			case Variant::TypeId::UINT64:
+			case Variant::TypeId::FLOAT:
+			case Variant::TypeId::DOUBLE:
+			{
+				// scalar concentration tolerance
+				// need to be converted to vector tolerance since speices might have various compartment sizes
+				double abstol = value.convert<double>();
+
+				for (int i = 0; i < mModel->getNumFloatingSpecies(); i++) {
+					// get the compartment volume of each species
+					int index = mModel->getCompartmentIndexForFloatingSpecies(i);
+					if (volumes[index] == 0) {
+						// when the compartment volume is 0, simply set the amount tolerance as the volume tolerance
```

Comment:
Should we throw an exception if a compartment's volume is 0?

Reply:
No need to throw an exception for now.